s3.tf

```diff
@@ -21,7 +24,8 @@ resource "aws_s3_bucket" "this" {
   server_side_encryption_configuration {
     rule {
       apply_server_side_encryption_by_default {
-        sse_algorithm = "aws:kms"
+        kms_master_key_id = local.s3_encryption_check[1]
```

Comment:
Not exactly sure, but it will encrypt nonetheless if kms is enabled or not or?

Reply:
It will fail if kms is not enabled.

Reply:
Looking at it from an (DAU) end-users point of view (and correct me if I am wrong here):

I want to create an S3 bucket with all `s3_*` options enabled, then during roll-out I get an error, that I don't have a KMS key. So therefore I must also explicitly enable `kms` when enabling S3, which is nowhere documented.

My proposal to make this more intuitive for Engineers is: 

* Remove the `kms_enabled` option
* Always automatically enable KMS whenever one of my specified services requires it.
* Later, when other services also make use of KMS, add this to the condition for auto-enabling KMS

What do u think about this workflow?

Reply:
I dislike it for the following reasons.
- The users of this module should read the documentation, certain features need to be documented and made clear in readme and examples. One can never have a totally dummy-proof solution, nor should we aim for one.
- `kms_enabled` enabled a KMS key, disabling the key will have an impact as the key won't be available for services to use. By giving it it's own configuration directive there is a clear definition of what is going to be created. 

By auto-creating a KMS key when another service might need it this is less clear that in the background a KMS key is created or dropped.

Reply:
I will also be initially fine, if the behaviour is documented well in the README and variable description. Nevertheless I would also like to have a second opinion here: @snovikov @ronny-panknin-flaconi @danvaida @czerasz-flaconi @vikkasyousaf

Reply:
I feel that skipping `kms_enabled` would be a good option:

- this way we guarantee that resources **always** use it
- I don't see the need of disabling KMS

We should document though that KMS related resources were created. I see output as part of the documentation

Reply:
It's a public module, so do we force to use KMS even when its not used ?

Reply:
This is a very specific module so why not? Alternatively we can create KMS resources only if the S3 bucket is enabled.

Reply:
I find enabling KMS always unnecessary, and enabling KMS in the background because it's utilized by something it needs problematic, for the reasons of transparency.

Reply:
On general I am on @maartenvanderhoef side and would like to keep control if KMS enabled or not, and especially because it is a public module.

Reply:
We should not really care wether or not this is a public module. First and foremost, we're coding it for our use case. If we can make it work for everyone, good, but we should not put this before our own use case.

It's 2019. Encryption has to be ubiquitous. KMS vs AES is a different story, but in-flight and at-rest encryption should not be an option anymore (see below).

Naturally, the list of resources created by the module (incl. the ones that the module's "API" is not exposing), should be part of the output.

The question is: if we enable KMS-encryption by default w/o giving the module users a choice, what happens to S3 buckets that need to contain public objects (i.e. some CDN origin)?
`GET` requests would fail with:
> Requests specifying Server Side Encryption with AWS KMS managed keys require AWS Signature Version 4.
If we have this use case already, sadly, I think it is legitimate enough to expose this encryption option --yes, we can set our preferred default (KMS encryption enabled).

How about back-porting existing buckets? Do we want to change their config or allow this to be a configurable?

I think that an acceptable compromise for now, would be to expose this variable that would trigger the handling of the CMK.

Reply:
Another option would be to enable/disable S3 bucket encyption (enabled by default). And whenever this is enabled, the KMS resources get created automatically in the background. The end-user does not really need to know about KMS, because he will never use it himself, or.

This would also allow backporting non-encrypted S3 buckets easily. WDYT?

Reply:
Backporting should not have entered this discussion. If we have legacy infra we can add that into terraform, one way or another but does not need to be back ported into this module. A module is being created to make it easy to deploy new infrastructure. You will never be able to mould all legacy into a module like this hence it's polluting the discussion here.

From what I get of the whole discussion is that the automatic encryption of resources should not be enabled. But we should instruct to enable it by default. 
Not because we don't want things encrypted automatically, but because it's more transparent this way.

With the instruction to enable it also comes the instruction to enable KMS which also needs to be explicitly enabled when it's used. If no objections I will make S3 more configurable and extend the documentation.

Reply:
The backporting was just an additional benefit. Nevertheless I do see it as an important aspect for the long run. For that we can have an offline discussion.

Apart from that what are you thoughts about having KMS created/removed in the background, as it will only be used by the services and policies itself and the user couldn't care less if this is done via KMS or AES.

Reply:
That is the big question if the user cares less about it, and if we care less about it. As long as this unclarity is there I think it's better to explicitly enable anything related to kms and encryption.

Reply:
I'd say that the user should not care about it, so handling it in the background is fine.